hack/update-codegen.sh

```diff
@@ -744,24 +844,286 @@ function codegen::informers() {
+#     An optional override for the leaf name of the generated "clientset" directory.
+#
+#   --versioned-name <string = "versioned">
+#     An optional override for the leaf name of the generated
+#     "<clientset>/versioned" directory.
+#
+#   --listers-name <string = "listers">
+#     An optional override for the leaf name of the generated "listers" directory.
+#
+#   --informers-name <string = "informers">
+#     An optional override for the leaf name of the generated "informers" directory.
+#
+#   --no-watch
+#     Disables generation of listers and informers for APIs which don't support
+#     WATCH.
+#
+function generate_client() {
+    local out_pkg_root
+    local in_pkg_root
+    local clientset_subdir="clientset"
+    local clientset_versioned_name="versioned"
+    local listers_subdir="listers"
+    local informers_subdir="informers"
+    local watchable="true"
+
+    while [ "$#" -gt 0 ]; do
+        case "$1" in
+            "--input_pkg_root")
+                in_pkg_root="$2"
+                shift 2
+                ;;
+            "--output_pkg_root")
+                out_pkg_root="$2"
+                shift 2
+                ;;
+            "--clientset-name")
+                clientset_subdir="$2"
+                shift 2
+                ;;
+            "--versioned-name")
+                clientset_versioned_name="$2"
+                shift 2
+                ;;
+            "--listers-name")
+                listers_subdir="$2"
+                shift 2
+                ;;
+            "--informers-name")
+                informers_subdir="$2"
+                shift 2
+                ;;
+            "--no-watch")
+                watchable="false"
+                shift
+                ;;
+            *)
+                echo "unknown argument: $1" >&2
+                return 1
+                ;;
+        esac
+    done
+
+    if [ -z "${in_pkg_root}" ]; then
+        echo "--input_pkg_root is required" >&2
+        return 1
+    fi
+    if [ -z "${out_pkg_root}" ]; then
+        echo "--output_pkg_root is required" >&2
+        return 1
+    fi
+
+    # The output pkg must exist at $root/$out_pkg_root because of gengo.
+    local root="staging/src"
+
+    local group_versions=()
+    local input_pkgs=()
+    while read -r file; do
+        dir="$(dirname "${file}")"
+        leaf="$(basename "${dir}")"
+        if grep -E -q --color=never '^v[0-9]+((alpha|beta)[0-9]+)?$' <<< "${leaf}"; then
+            input_pkgs+=("${dir#"${root}/"}") # drop "$root"
+
+            dir2="$(dirname "${dir}")"
+            leaf2="$(basename "${dir2}")"
+            group_versions+=("${leaf2}/${leaf}")
+        fi
+    done < <(git ls-files -cmo --exclude-standard -z \
+        ":(glob)${root}/${in_pkg_root}"/'**/types.go' \
+        | xargs -0 grep -l --color=never '+genclient' \
+        | sort -u)
+
+    if [ "${#group_versions[@]}" == 0 ]; then
+        return 0
+    fi
+
+    GO111MODULE=on GOPROXY=off \
```

Comment:
GOPROXY=off is surprising... what's up with that (same question applies elsewhere)?

Reply:
That's a pattern copied from the original source.  These probably ALL should move from `go install` to `hack/make-rules/build.sh`.  Can be a pre-factoring.

Reply:
https://github.com/kubernetes/kubernetes/pull/115171